Source/CombatExtended/CombatExtended/ShellingUtility.cs

```diff
@@ -0,0 +1,145 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using RimWorld;
+using RimWorld.Planet;
+using UnityEngine;
+using Verse;
+using CombatExtended.WorldObjects;
+
+namespace CombatExtended
+{
+    public static class ShellingUtility
+    {
+        private static Map map;
+        private static ThingDef shellDef;
+        private static ProjectilePropertiesCE props;
+        private static DamageDef projectileDamageDef;
+
+        public static IntVec3 FindRandomImpactCell(Map map, ThingDef shellDef = null)
+        {
+            ShellingUtility.map = map;
+            ShellingUtility.shellDef = shellDef;
+            IntVec3 result = IntVec3.Invalid;
+            if (map.IsPlayerHome)
+            {
+                props = (ProjectilePropertiesCE)shellDef.projectile;
+                projectileDamageDef = props.damageDef;
+                Stopwatch stopwatch = new Stopwatch();
+                stopwatch.Start();
+                int i = 5;
+                while (i-- >= 0 && stopwatch.ElapsedTicks / Stopwatch.Frequency < 0.01f)
+                {
+                    result = GetRandomPlayerHomeImpactCell();
+                    if (result.IsValid)
+                    {
+                        return result;
+                    }
+                }
+                stopwatch.Stop();
+            }
+            if (!result.IsValid)
+            {
+                result = GetRandomImpactCell();
+            }
+            map = null;
+            return result;
+        }
+
+        private static IntVec3 GetRandomPlayerHomeImpactCell()
+        {
+            if (projectileDamageDef == CE_DamageDefOf.Bomb)
+            {
+                if (Rand.Chance(props.damageAmountBase / 5000f) && map.areaManager.Home.innerGrid.trueCountInt > 0) // make more powerful shells land near the player base
```

Comment:
This could probably use some refinement at some point.  I've zoned large patches of home zone under a mountain, just to make sure that anyone restricted to home can reach where ever I decide to mine.